src/Aspire.Dashboard/Model/DashboardClient.cs

```diff
@@ -0,0 +1,355 @@
+
+                    lock (_lock)
+                    {
+                        // We received a message, which means we are connected. Clear the error count.
+                        errorCount = 0;
+
+                        if (response.KindCase == WatchResourcesUpdate.KindOneofCase.InitialData)
+                        {
+                            // Populate our map using the initial data.
+                            _resourceByName.Clear();
+
+                            // TODO send a "clear" event via outgoing channels, in case consumers have extra items to be removed
+
+                            foreach (var resource in response.InitialData.Resources)
+                            {
+                                // Add to map.
+                                var viewModel = resource.ToViewModel();
+                                _resourceByName[resource.Name] = viewModel;
+
+                                // Send this update to any subscribers too.
+                                changes ??= [];
+                                changes.Add(new(ResourceViewModelChangeType.Upsert, viewModel));
+                            }
+                        }
+                        else if (response.KindCase == WatchResourcesUpdate.KindOneofCase.Changes)
+                        {
+                            // Apply changes to the model.
+                            foreach (var change in response.Changes.Value)
+                            {
+                                changes ??= [];
+
+                                if (change.KindCase == WatchResourcesChange.KindOneofCase.Upsert)
+                                {
+                                    // Upsert (i.e. add or replace)
+                                    var viewModel = change.Upsert.ToViewModel();
+                                    _resourceByName[change.Upsert.Name] = viewModel;
+                                    changes.Add(new(ResourceViewModelChangeType.Upsert, viewModel));
+                                }
+                                else if (change.KindCase == WatchResourcesChange.KindOneofCase.Delete)
+                                {
+                                    // Remove
+                                    if (_resourceByName.Remove(change.Delete.ResourceName, out var removed))
+                                    {
+                                        changes.Add(new(ResourceViewModelChangeType.Delete, removed));
+                                    }
+                                    else
+                                    {
+                                        Debug.Fail("Attempt to remove an unknown resource view model.");
+                                    }
+                                }
+                                else
+                                {
+                                    throw new FormatException($"Unexpected {nameof(WatchResourcesChange)} kind: {change.KindCase}");
```

Comment:
nit: FormatException isn't the exception I'd expect here. I would expect ArgumentOutOfRangeException or InvalidOperationException